cpp/src/arrow/compute/kernels/vector_sort_test.cc

```diff
@@ -2115,6 +2113,100 @@ INSTANTIATE_TEST_SUITE_P(AllNull, TestTableSortIndicesRandom,
                          testing::Combine(first_sort_keys, num_sort_keys,
                                           testing::Values(1.0)));
 
+class TestNestedSortIndices : public ::testing::Test {
+ protected:
+  static std::shared_ptr<Array> GetArray() {
+    auto struct_type =
+        struct_({field("a", struct_({field("a", uint8()), field("b", uint32())})),
+                 field("b", int32())});
+    auto struct_array = checked_pointer_cast<StructArray>(
+        ArrayFromJSON(struct_type,
+                      R"([{"a": {"a": 5,    "b": null}, "b": 8   },
+                          {"a": {"a": null, "b": 7   }, "b": 3   },
+                          {"a": {"a": null, "b": 9   }, "b": 3   },
+                          {"a": {"a": 2,    "b": 4   }, "b": 6   },
+                          {"a": {"a": 5,    "b": 1   }, "b": null},
+                          {"a": {"a": 3,    "b": null}, "b": 2   },
+                          {"a": {"a": 2,    "b": 3   }, "b": 0   },
+                          {"a": {"a": 2,    "b": 4   }, "b": 1   },
+                          {"a": {"a": null, "b": 7   }, "b": null}])"));
```

Comment:
Hmm... perhaps we can instead have a struct-of-struct-of-struct? This would help exercise the case where a sort key points to a struct-inside-struct.